src/persist/src/indexed/mod.rs

```diff
@@ -131,6 +131,52 @@ impl<U: Buffer, L: Blob> Indexed<U, L> {
         Ok(indexed)
     }
 
+    /// Revert the in-memory state back to a previously serialized version.
+    ///
+    /// Used to keep the in-memory and durably stored data structures consistent
+    /// in the presence of errors.
+    fn restore(&mut self, meta: BlobMeta) {
```

Comment:
I feel like there should be a way for this to share most of its code with new, so restore becomes something like `*self = new_indexed_from_meta(self.blob, meta)`. The next_blob_id stuff is an issue, but they don't really have to be incrementing, so we could find the max of any of them and pass that +1 in as a parameter to new_indexed_from_meta as well and use it for all of them? Thoughts?

Reply:
I think thats correct - let me see if I can whip something up. I think you have to be careful to hold onto the existing listeners which might be why I left this this way last night.

Reply:
Ah the listeners too, that's a good point. I'm also fine going with this approach for now given that we might have to change it anyway when we actually start batching stuff (plus at some point we should maybe do minimal rollbacks?)

Reply:
left a todo